File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] arr) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] pre : arr) {
            adj.get(pre[1]).add(pre[0]);
        }

        int[] vis = new int[numCourses];
        int[] pathvis = new int[numCourses];
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < numCourses; i++) {
            if (vis[i] == 0) {
                if (dfs(i, adj, vis, pathvis, stack)) {
                    // cycle detected → no valid order
                    return new int[0];
                }
            }
        }

        int[] order = new int[numCourses];
        int idx = 0;
        while (!stack.isEmpty()) {
            order[idx++] = stack.pop();
        }
        return order;
    }

    static boolean dfs(int node, ArrayList<ArrayList<Integer>> adj,
                       int[] vis, int[] pathvis, Stack<Integer> stack) {
        vis[node] = 1;
        pathvis[node] = 1;

        for (int nei : adj.get(node)) {
            if (vis[nei] == 0) {
                if (dfs(nei, adj, vis, pathvis, stack)) return true;
            } else if (pathvis[nei] == 1) {
                return true; // cycle found
            }
        }

        pathvis[node] = 0;
        stack.push(node);
        return false;
    }
}

File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    bool dfs(int curr,vector<int>& visited,vector<vector<int>>& adj,stack<int>& st){
        visited[curr]=1;
        for(auto it:adj[curr]){
            if(visited[it]==0){
                if(!dfs(it,visited,adj,st)){
                    return false;
                }
            }else if(visited[it]==1){
                return false;
            }
        }
        visited[curr]=2;
        st.push(curr);
        return true;
    }
    vector<int> findOrder(int num, vector<vector<int>>& prerequisites) {
        int n=prerequisites.size();
        vector<vector<int>> adj(num);
        for(int i=0;i<n;i++){
            adj[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }
        
        stack<int> st;
        vector<int> visited(num,0);
        for(int i=0;i<num;i++){
            if(visited[i]==0){
                if(!dfs(i,visited,adj,st)){
                    return {};
                }
            }
        }
        vector<int> ans;
        while(!st.empty()){
            ans.push_back(st.top());
           st.pop();
        }
        return ans;
    }
};